xonsh/xoreutils/cat.py

```diff
@@ -0,0 +1,104 @@
+import os
+import sys
+
+from xonsh.xoreutils.util import arg_handler
+
+def _cat_single_file(opts, fname, stdin, out, err, line_count=1):
+    if fname == '-':
+        f = stdin
+    elif os.path.isdir(fname):
+        print("cat: {}: Is a directory.".format(fname), file=err)
+        return True, line_count
+    elif not os.path.exists(fname):
+        print("cat: No such file or directory: {}".format(fname), file=err)
+        return True, line_count
+    else:
+        f = open(fname, 'rb')
```

Comment:
Don't use `os.path.*`. Instead catch exceptions here. Between your "exists" call and your "open" call, the file may have been removed (or a symlink could have beene swung to point to a directory, etc etc). These error messages should come as a result of an exception trying to open the file.

Compare the behaviour of this:
```
try:
    open(fname, 'rb')
except IOError as err:
    print("cat: {}: {}".format(err.strerror, fname), file=err)
```